### PR TITLE
Fix #1475

### LINKED ISSUE #1475: Memory leaks in Zephir
```
==32320== 
==32320== HEAP SUMMARY:
==32320==     in use at exit: 22,383 bytes in 259 blocks
==32320==   total heap usage: 313,020 allocs, 312,761 frees, 66,395,022 bytes allocated
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 13 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x106CDEA0: zim_Test_Sort_quick (sort.zep.c:95)
==32320==    by 0x848517: zend_call_function (zend_execute_API.c:885)
==32320==    by 0x105F7B1A: zephir_call_user_function (fcall.c:303)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x106CDEA0: zim_Test_Sort_quick (sort.zep.c:95)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 14 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x10670D83: zim_Test_Mcall_testCall7 (mcall.zep.c:242)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 15 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x10670E6C: zim_Test_Mcall_testCall7 (mcall.zep.c:244)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 16 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x106710BD: zim_Test_Mcall_testCall8 (mcall.zep.c:268)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 17 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x106711A6: zim_Test_Mcall_testCall8 (mcall.zep.c:270)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 16 bytes in 1 blocks are definitely lost in loss record 18 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x105F7BA3: zephir_call_user_function (fcall.c:323)
==32320==    by 0x105F8330: zephir_call_class_method_aparams (fcall.c:509)
==32320==    by 0x106A9A4A: zim_Test_Pregmatch_testMatchAllInZep (pregmatch.zep.c:320)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 56 bytes in 1 blocks are definitely lost in loss record 39 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x865702: _array_init (zend_API.c:1061)
==32320==    by 0x106A7D6F: zim_Test_Pregmatch_testPregMatchAll (pregmatch.zep.c:119)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320== 
==32320== 96 bytes in 2 blocks are definitely lost in loss record 40 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x8A7FFB: zend_objects_new (zend_objects.c:162)
==32320==    by 0x8666C9: _object_and_properties_init (zend_API.c:1301)
==32320==    by 0x866786: _object_init_ex (zend_API.c:1316)
==32320==    by 0x8C6DB4: ZEND_NEW_SPEC_CONST_HANDLER (zend_vm_execute.h:3342)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320== 
==32320== 504 (56 direct, 448 indirect) bytes in 1 blocks are definitely lost in loss record 58 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x865702: _array_init (zend_API.c:1061)
==32320==    by 0x7DD61F: php_build_argv (php_variables.c:537)
==32320==    by 0x7DDE71: php_hash_environment (php_variables.c:666)
==32320==    by 0x7C71B3: php_request_startup (main.c:1628)
==32320==    by 0x9283A0: do_cli (php_cli.c:945)
==32320==    by 0x929724: main (php_cli.c:1344)
==32320==
==32320== 736 (584 direct, 152 indirect) bytes in 3 blocks are definitely lost in loss record 60 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x8A7FFB: zend_objects_new (zend_objects.c:162)
==32320==    by 0x8666C9: _object_and_properties_init (zend_API.c:1301)
==32320==    by 0x866786: _object_init_ex (zend_API.c:1316)
==32320==    by 0x8C6DB4: ZEND_NEW_SPEC_CONST_HANDLER (zend_vm_execute.h:3342)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8BF4BC: zend_execute (zend_vm_execute.h:458)
==32320==    by 0x8605FB: zend_execute_scripts (zend.c:1427)
==32320==    by 0x7C8D9F: php_execute_script (main.c:2494)
==32320==    by 0x928556: do_cli (php_cli.c:974)
==32320==    by 0x929724: main (php_cli.c:1344)
==32320==
==32320== 1,924 (208 direct, 1,716 indirect) bytes in 1 blocks are definitely lost in loss record 67 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x8040A3: compile_file (zend_language_scanner.l:596)
==32320==    by 0x58805F: phar_compile_file (phar.c:3322)
==32320==    by 0x86058D: zend_execute_scripts (zend.c:1421)
==32320==    by 0x7C8D9F: php_execute_script (main.c:2494)
==32320==    by 0x928556: do_cli (php_cli.c:974)
==32320==    by 0x929724: main (php_cli.c:1344)
==32320==
==32320== 6,608 bytes in 118 blocks are definitely lost in loss record 69 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x865702: _array_init (zend_API.c:1061)
==32320==    by 0x2F6F51: php_pcre_match_impl (php_pcre.c:697)
==32320==    by 0x105F6208: zephir_preg_match (string.c:1082)
==32320==    by 0x106BD268: zim_Test_Router_handle (router.zep.c:634)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==
==32320== 6,968 (448 direct, 6,520 indirect) bytes in 8 blocks are definitely lost in loss record 70 of 70
==32320==    at 0x4C2CB3F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32320==    by 0x82AE2E: _emalloc (zend_alloc.c:2446)
==32320==    by 0x865702: _array_init (zend_API.c:1061)
==32320==    by 0x2F6F51: php_pcre_match_impl (php_pcre.c:697)
==32320==    by 0x105F6208: zephir_preg_match (string.c:1082)
==32320==    by 0x106B6146: zim_Test_RegexDNA_process (regexdna.zep.c:192)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==    by 0x8483F1: zend_call_function (zend_execute_API.c:866)
==32320==    by 0x5AF943: zim_reflection_method_invokeArgs (php_reflection.c:3348)
==32320==    by 0x8C052B: ZEND_DO_FCALL_SPEC_HANDLER (zend_vm_execute.h:842)
==32320==    by 0x8BF3AB: execute_ex (zend_vm_execute.h:414)
==32320==
==32320== LEAK SUMMARY:
==32320==    definitely lost: 8,152 bytes in 140 blocks
==32320==    indirectly lost: 8,836 bytes in 79 blocks
==32320==      possibly lost: 0 bytes in 0 blocks
==32320==    still reachable: 5,395 bytes in 40 blocks
==32320==         suppressed: 0 bytes in 0 blocks
```

### PR DESCRIPTION
**Pull request by sjinks**
WIP, do not merge

Depends on:
* [x] #1480 (fixed in #1515)
* [x] #1481 (fixed in #1515)
* [x] #1482 (fixed in #1515)
* [x] #1483 (fixed in #1515)
* [x] #1486 (fixed in #1513)
* [ ] #1487
* [x] #1516 (fixed in #1517, #1518)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -61,7 +61,7 @@ before_script:
 
 script:
   - vendor/bin/phpcs --standard=PSR2 --report=emacs --extensions=php --warning-severity=0 Library/ unit-tests/Extension/ unit-tests/Zephir/
-  - valgrind --read-var-info=yes --error-exitcode=1 --fullpath-after= --track-origins=yes --leak-check=full ./vendor/bin/phpunit -c phpunit.xml.dist --debug unit-tests/
+  - USE_ZEND_ALLOC=0 ZEND_DONT_UNLOAD_MODULES=1 valgrind --read-var-info=yes --error-exitcode=1 --fullpath-after= --track-origins=yes --leak-check=full --suppressions=unit-tests/ci/zephir.3.7.0.sup $(phpenv which php) ./vendor/bin/phpunit -c phpunit.xml.dist --debug unit-tests/
```
Comment (sergeyklay) [inline .travis.yml]: Can we use this globally ?

```yaml
env:
  global:
    - LD_LIBRARY_PATH="$TRAVIS_BUILD_DIR/build/lib":$LD_LIBRARY_PATH
    - LIBRARY_PATH="$TRAVIS_BUILD_DIR/build/lib":$LIBRARY_PATH
    - C_INCLUDE_PATH="$TRAVIS_BUILD_DIR/build/include"
    - CFLAGS="-g3 -O0 -Wall -fvisibility=hidden"
    - ZEND_DONT_UNLOAD_MODULES=1
```

Comment (sjinks) [reply .travis.yml]: Yes.

**Comment by sjinks** [comment]
See #1480

File: Library/Backends/ZendEngine2/Backend.php [hunk 2]
```diff
@@ -269,12 +269,38 @@ public function onPostCompile($method, CompilationContext $context)
...
+                    function ($matches) use ($cleanup) {
+                        return $cleanup . $matches[1];
+                    },
+                    $code
+                );
+                $codePrinter->clear();
+                $codePrinter->output($code);
+            }
+        }
     }
```
Comment (sjinks) [inline Library/Backends/ZendEngine2/Backend.php]: @sergeyklay could you please review this?

File: Library/Backends/ZendEngine2/Backend.php [hunk 3]
```diff
@@ -269,12 +269,34 @@ public function onPostCompile($method, CompilationContext $context)
...
```
Comment (sjinks) [inline Library/Backends/ZendEngine2/Backend.php]: @sergeyklay could you please review this?

Comment (sjinks) [reply Library/Backends/ZendEngine2/Backend.php]: The entire block :-)

File: Library/Backends/ZendEngine2/Backend.php [hunk 4]
```diff
@@ -269,12 +269,34 @@ public function onPostCompile($method, CompilationContext $context)
             $codePrinter->output('}');
             $codePrinter->decreaseLevel();
         }
+        else {
+            $cleanup = '';
+            foreach ($context->symbolTable->getVariables() as $variable) {
+                if ($variable->getName() != 'this_ptr' && $variable->getName() != 'return_value' && $variable->getName() != 'return_value_ptr') {
```
Comment (Jurigag) [inline Library/Backends/ZendEngine2/Backend.php]:
```php
if (!in_array($variable->getName(), ['this_ptr', 'return_value', 'return_value_ptr'])) {
```

?

File: Library/Backends/ZendEngine2/Backend.php [hunk 5]
```diff
@@ -253,17 +264,39 @@ public function onPostCompile($method, CompilationContext $context)
...
-            $codePrinter->output('return Z_OBJVAL_P(this_ptr);');
+            $codePrinter->output('return retval;');
             $codePrinter->decreaseLevel();
             $codePrinter->output('}');
             $codePrinter->decreaseLevel();
         }
+        else {
+            $cleanup = '';
+            foreach ($context->symbolTable->getVariables() as $variable) {
+                if ($variable->getName() != 'this_ptr' && $variable->getName() != 'return_value' && $variable->getName() != 'return_value_ptr') {
```
Comment (sergeyklay) [inline Library/Backends/ZendEngine2/Backend.php]: Actually I like more something like:

```php
$cleanup  = '';
$callback = function($variable) use (&$cleanup) {
    /// ...
};

array_map($callback, $context->symbolTable->getVariables());
```

File: Library/Backends/ZendEngine2/Backend.php [hunk 6]
```diff
@@ -253,17 +264,39 @@ public function onPostCompile($method, CompilationContext $context)
...
+                    if ($variable->isComplex() && $variable->mustInitNull() && $variable->isLocalOnly()) {
+                        $cleanup .= "\t" . 'zval_dtor(&' . $variable->getName() . ');' . PHP_EOL;
```
Comment (sergeyklay) [inline Library/Backends/ZendEngine2/Backend.php]: Could we incapsulate this into something like 
```php
$variable->haveToBeCleaned();
```

I'd like to cover this by tests in near future and have an ability to mock this

File: Library/Variable.php [hunk 7]
```diff
@@ -980,6 +980,17 @@ public function observeOrNullifyVariant(CompilationContext $compilationContext)
...
 
     /**
+     * Whether the variable is of a complex type
+     *
+     * @return boolean
+     */
+    public function isComplex()
+    {
+        $type = $this->type;
+        return ($type == 'variable' || $type == 'string' || $type == 'array' || $type == 'resource' || $type == 'callable' || $type == 'object');
```
Comment (sergeyklay) [inline Library/Variable.php]:
```php
protected $complexVariables = [
    'variable' => true,
    'string'   => true,
    // ..
];

public function isComplex()
{
    return isset($this->complexVariables[$this->type]);
}

File: Library/Backends/ZendEngine2/Backend.php [hunk 8]
```diff
@@ -242,8 +242,19 @@ public function onPreCompile($method, CompilationContext $context)
...
+            $codePrinter->output('zend_object_value retval;');
+            $codePrinter->outputBlankLine();
+            $codePrinter->output('zend_object_std_init(obj, class_type TSRMLS_CC);');
+            $codePrinter->output('object_properties_init(obj, class_type);');
+            $codePrinter->output('retval.handle   = zend_objects_store_put(obj, (zend_objects_store_dtor_t)zend_objects_destroy_object, zephir_free_object_storage, NULL TSRMLS_CC);');
+            $codePrinter->output('retval.handlers = zend_get_std_object_handlers();');
+            $codePrinter->outputBlankLine();
+            $codePrinter->output('Z_TYPE(zthis)   = IS_OBJECT;');
+            $codePrinter->output('Z_OBJVAL(zthis) = retval;');
+            $codePrinter->outputBlankLine();
```
Comment (sergeyklay) [inline Library/Backends/ZendEngine2/Backend.php]: Could we use an template file for all this?

File: kernels/ZendEngine2/exception.c [hunk 9]
```diff
@@ -49,21 +49,21 @@ void zephir_throw_exception_debug(zval *object, const char *file, zend_uint line
...
-		object_copy = object;
+		zval *arg = object;
 		ALLOC_INIT_ZVAL(object);
 		object_init_ex(object, zend_exception_get_default(TSRMLS_C));
-		ZEPHIR_CALL_METHOD(NULL, object, "__construct", NULL, 0, object_copy);
+		ZEPHIR_CALL_METHOD(NULL, object, "__construct", NULL, 0, arg);
+	}
+	else {
+		Z_ADDREF_P(object);
 	}
```
Comment (sergeyklay) [inline kernels/ZendEngine2/exception.c]: could we introduce a small factory function to incapsulate this logic and/or increment refcount

File: kernels/ZendEngine2/globals.h [hunk 10]
```diff
@@ -90,24 +86,6 @@ typedef zend_function zephir_fcall_cache_entry;
...
-	Z_SET_REFCOUNT_P(z, 1); \
-	Z_UNSET_ISREF_P(z);
-#endif
-
-#ifndef ZVAL_COPY_VALUE
-#define ZVAL_COPY_VALUE(z, v) \
-	(z)->value = (v)->value; \
-	Z_TYPE_P(z) = Z_TYPE_P(v);
-#endif
-
```
Comment (sergeyklay) [inline kernels/ZendEngine2/globals.h]: Thank you!

File: kernels/ZendEngine2/main.h [hunk 11]
```diff
@@ -50,19 +50,6 @@
...
-  (z)->value = (v)->value;\
-  Z_TYPE_P(z) = Z_TYPE_P(v);
-#endif
-
-#ifndef INIT_PZVAL_COPY
- #define INIT_PZVAL_COPY(z, v) ZVAL_COPY_VALUE(z, v);\
-  Z_SET_REFCOUNT_P(z, 1);\
-  Z_UNSET_ISREF_P(z);
-#endif
-
```
Comment (sergeyklay) [inline kernels/ZendEngine2/main.h]: 👍

File: kernels/ZendEngine2/exception.c (hunk 9 above)
Comment (sergeyklay) [reply kernels/ZendEngine2/exception.c]: I mean something like (not sure about macros name)

```c
#define CREATE_OBJECT_OR_ADD_REF(zval_ptr) \
    if (Z_TYPE_P(zval_ptr) != IS_OBJECT) { \
        zval *arg = object; \
        ALLOC_INIT_ZVAL(object); \
        object_init_ex(object, zend_exception_get_default(TSRMLS_C)); \
        ZEPHIR_CALL_METHOD(NULL, object, "__construct", NULL, 0, arg); \
    } else { \
        Z_ADDREF_P(zval_ptr); \
    }
```

**Comment by sergeyklay** [comment]
@sjinks I'm sorry for conflict. Pushed to the master by mistake  https://github.com/phalcon/zephir/issues/1475#issuecomment-292666931

File: Library/Backends/ZendEngine2/Backend.php (hunk 4 above)
Comment (sjinks) [reply Library/Backends/ZendEngine2/Backend.php]: This gathers all used variables (the code is from https://github.com/phalcon/zephir/blob/master/Library/ClassMethod.php#L1999), then retrieves only those which are initialized with `zval_used_for_init` (https://github.com/phalcon/zephir/blob/master/Library/Backends/ZendEngine2/Backend.php#L268).

File: Library/Backends/ZendEngine2/Backend.php (hunk 6 above)
Comment (sjinks) [reply Library/Backends/ZendEngine2/Backend.php]: The only issue is that this is ZE2-specific method :-(

File: kernels/ZendEngine2/exception.c (hunk 9 above)
Comment (sjinks) [reply kernels/ZendEngine2/exception.c]: This is specific only to this method

File: Library/Variable.php (hunk 7 above)
Comment (sjinks) [inline Library/Variable.php]: +

File: Library/Backends/ZendEngine2/Backend.php (hunk 6 above)
Comment (sergeyklay) [reply Library/Backends/ZendEngine2/Backend.php]: Well, never mind then 👍

File: Library/Backends/ZendEngine2/Backend.php (hunk 5 above)
Comment (sjinks) [reply Library/Backends/ZendEngine2/Backend.php]: The current implementation is much like a hack :-(

There is no `cleanup()` method or something like this, therefore we have to (ab)use `onPostCompile()`. But the issue is that in `postCompile()` the entire method has already been generated, and we have to deal with the code in the text form.

The current logic is as follows:
* we get the list of all used variables and filter only the local ones;
* for every local variable we add a call to `zval_dtor()`;
* all these calls are saved into a string;
* then we get the text code from the code printer and insert out string with `zval_dtor()`s right before the `RETURN_xxx` statement.

I know that this is not an elegant solution, but with the current implementation of the compiler I was unable to come up with something more beautiful :-(

**Comment by sjinks** [comment]
Looking at the macros, I see that f70ca78 is not a good solution :-(

  * if there is an exception in the middle of the method, there will be a leak
  * if there is a `return` in the middle of the method, there will be a leak

It looks like I will have to mess with the memory manager to add another kind of observed variables for which `zval_dtor` will need to be called, not `zval_ptr_dtor`. Sigh.

File: .travis.yml [hunk 12]
```diff
@@ -47,6 +49,7 @@ cache:
     - $HOME/.composer/cache
 
 install:
+  - phpenv config-rm xdebug.ini
```
Comment (sergeyklay) [inline .travis.yml]: Could you add `- phpenv config-rm xdebug.ini || true`

File: Library/Backends/ZendEngine2/Backend.php (hunk 8 above)
Comment (sjinks) [reply Library/Backends/ZendEngine2/Backend.php]: You mean, a macro?

Comment (sergeyklay) [reply Library/Backends/ZendEngine2/Backend.php]: No, I suggest to use something like `templates/ZendEngine2/zephir_init_properties.tpl`. I'm still not sure. Just wondering your opinion

File: Library/Backends/ZendEngine2/Backend.php [hunk 13]
```diff
@@ -347,7 +347,6 @@ public function initializeVariableDefaults($variables, CompilationContext $compi
 
         /* Initialize default values in dynamic variables */
         foreach ($variables as $variable) {
-
```
Comment (sjinks) [inline Library/Backends/ZendEngine2/Backend.php]: Thanks

**Comment by sjinks** [comment]
Rebased against the most current master

**Comment by sergeyklay** [comment]
@sjinks Should we submit this to the `development` branch?

**Pull request closed without merging**